pkg/fleet-manager/application/rollout_helper.go

```diff
@@ -0,0 +1,459 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package application
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	flaggerv1b1 "github.com/fluxcd/flagger/pkg/apis/flagger/v1beta1"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	applicationapi "kurator.dev/kurator/pkg/apis/apps/v1alpha1"
+	fleetapi "kurator.dev/kurator/pkg/apis/fleet/v1alpha1"
+	fleetmanager "kurator.dev/kurator/pkg/fleet-manager"
+	"kurator.dev/kurator/pkg/fleet-manager/plugin"
+)
+
+const (
+	RolloutLabel = "kurator.dev/rollout"
+)
+
+func (a *ApplicationManager) fetchRolloutClusters(ctx context.Context,
+	app *applicationapi.Application,
+	kubeClient client.Client,
+	originFleet *fleetapi.Fleet,
+	syncPolicy *applicationapi.ApplicationSyncPolicy) (map[fleetmanager.ClusterKey]*fleetmanager.FleetCluster, error) {
+
+	log := ctrl.LoggerFrom(ctx)
+	destination := getPolicyDestination(app, syncPolicy)
+	ClusterInterfaceList, result, err := a.fetchFleetClusterList(ctx, originFleet, destination.ClusterSelector)
+	if err != nil || result.RequeueAfter > 0 {
+		return nil, err
+	}
+
+	fleetclusters := make(map[fleetmanager.ClusterKey]*fleetmanager.FleetCluster, len(ClusterInterfaceList))
+	for _, cluster := range ClusterInterfaceList {
+		kclient, err := fleetmanager.ClientForCluster(kubeClient, originFleet.Namespace, cluster)
+		if err != nil {
+			return nil, err
+		}
+
+		kind := cluster.GetObject().GetObjectKind().GroupVersionKind().Kind
+		fleetclusters[fleetmanager.ClusterKey{Kind: kind, Name: cluster.GetObject().GetName()}] = &fleetmanager.FleetCluster{
+			Secret:    cluster.GetSecretName(),
+			SecretKey: cluster.GetSecretKey(),
+			Client:    kclient,
+		}
+	}
+	log.Info("Successful to fetch destination clusters for Rollout")
+	return fleetclusters, nil
+}
+
+func (a *ApplicationManager) syncRolloutPolicyForCluster(ctx context.Context,
+	rolloutPolicy *applicationapi.RolloutConfig,
+	destinationClusters map[fleetmanager.ClusterKey]*fleetmanager.FleetCluster,
+	policyName string) (ctrl.Result, error) {
+
+	log := ctrl.LoggerFrom(ctx)
+
+	// check some conflicting fields
+	if result, err := checkRolloutConflict(ctx, rolloutPolicy); err != nil {
+		return result, fmt.Errorf("failed to start rollout because: %s", err)
+	}
+
+	label := map[string]string{
+		RolloutLabel: policyName,
+	}
+
+	for clusterKey, fleetCluster := range destinationClusters {
+		// If the workload is a deployment, get its information to create the HPA.
+		if rolloutPolicy.Workload.Kind == "Deployment" {
```

Comment:
how about other kind?

Reply:
Other kind is DeamonSet, which doesn't need HPA